### DISCUSSION TITLE
Get CLOB from Stored Procedure

### DISCUSSION THREAD

**Discussion by osharko** [godror/godror#217] [Q&A] [answered]
Hi, i have a procedure who gave in output a clob:
```
procedure GET_CSV_FILE(P_CODE  in VARCHAR2, P_SEPARATOR  in VARCHAR2, P _CSV_FILE OUT CLOB)
```

Based on P_Code, it will return a string with differnt length. I'm able to store some of them to a string, but not the longest.

    "ORA-22835: Buffer too small for CLOB to CHAR or BLOB to RAW conversion (actual: 247005, maximum: 4000)"

This is the code i'm using:

      var csv string
      stmt.ExecContext(ctx,
        sql.Named("P_CODE", "GGS_LOP"),
        sql.Named("P_SEPARATOR", ""),
        sql.Named("P_CSV_FILE", sql.Out{Dest: &csv}),
      )

I also tried adding the  godror.LobAsReader() as args, using a bytes/strings Reader, or a []bytes but all of this type are refused.

How should I retrieve this long CLOB from the procedure?

Thank you

**Comment by tgulacsi** [comment, marked as answer]
Is this Oracle DB 11 ?

```
var csv godror.Lob
  stmt.ExecContext(ctx,
godror.LobAsReader() ,
    sql.Named("P_CODE", "GGS_LOP"),
    sql.Named("P_SEPARATOR", ""),
    sql.Named("P_CSV_FILE", sql.Out{Dest: &csv}),
  )
...
io.Copy(os.Stdout, csv.Reader)
```

**Reply by osharko** [reply]
Yes, this is still the 11. This is the result:
`"ORA-06553: PLS-306: wrong number or types of arguments in call to 'GET_CSV_FILE'"`

**Reply by osharko** [reply]
Same on Oracle 19

**Reply by tgulacsi** [reply]
Without sql.Named? Usually I use positional arguments.

**Reply by osharko** [reply]
I made the following: icsv := godror.Lob{IsClob: true}
With this initialization specifing the IsClob, it worked.
Just to be sure, is the code right?

```
var csv string
icsv := godror.Lob {IsClob: true}
  stmt.ExecContext(ctx,
godror.LobAsReader() ,
    sql.Named("P_CODE", "GGS_LOP"),
    sql.Named("P_SEPARATOR", ""),
    sql.Named("P_CSV_FILE", sql.Out{Dest: &csv}),
  )
...
buf := new(strings.Builder)
io.Copy(buf, icsv.Reader)
csv = buf.String()
```

**Reply by osharko** [reply]
Using IsClob: true, works even without the godror.LobAsReader()...

**Reply by tgulacsi** [reply]
LobAsReader is only for SELECT statements where you don't declare what type you want.

Your code is correct, and yes, you have to specify IsClob. Should it be better documented?

**Reply by osharko** [reply]
I think that a test case, linked into the documentation, could be very usefull.
We could think to do a kind of z-example.go where we link the example code to the docs.
If you're interested, I could make some example for the store procedures part, the only one I used.
